src/api/c/blas.cpp

```diff
@@ -145,23 +147,25 @@ af_err af_matmul(af_array *out, const af_array lhs, const af_array rhs,
         }
 
         TYPE_ASSERT(lhs_type == rhs_type);
-        af_array output = 0;
+        if (*out == 0) {
+            const dim4 outDims(0, 0, 0, 0);
+            *out = createHandle(outDims, lhs_type);
+        }
```

Comment:
```
af_array out; // This is not zero.
af_matmul(&out, lhs, rhs, opts);
```
This is now broken. This could break python and other language wrappers because the pointers are not explicitly set to NULL.

If you want to support a way to reuse existing array for matmul, write a new function. Do not have hidden behavior this way.

Reply:
Thank you for pointing this out. This may be beside the fact, but I remember @umar456 has also said that this needs an assertion that `out` is null indeed.

In any case, @9prady9 and @umar456, thoughts on this? It's a valid point that pointers do need to be always initialized after all, and the question then now becomes: does the responsibility of initializing the af_array mainly to the user? Or to the library? Based on the [previous code](https://github.com/arrayfire/arrayfire/blob/6c28d07218045f8ab997a6b455db24a8786a3f5b/src/api/c/blas.cpp#L85) before this PR, it seems like the responsibility has been with the library for a long time now, and thus this is the expected behavior among arrayfire users (as it is still currently with other functions). Is there any other way to determine if `out` is not a pre-allocated/existing array (that covers both null and uninitialized pointers)?

Reply:
I vote for AF to init/create/allocate if the out is null/empty.